### ISSUE TITLE
Serilization of np integers and floats

### ISSUE THREAD

**Issue by janvanrijn** [openml/openml-python#304]
Seems like the sklearn to flow serialization does not like it when a np.int64 is passed through:
```
  File "/home/vanrijn/projects/openml-python/openml/flows/sklearn_converter.py", line 81, in sklearn_to_flow
    raise TypeError(o, type(o))
TypeError: (2, <class 'numpy.int64'>)
```
Apparently, that case is not caught by
`elif isinstance(o, (bool, int, float, six.string_types)) or o is None:`

**Comment by mfeurer** [comment]
Correct, but why can't you use normal integers?

**Comment by PGijsbers** [comment]
I ran into this as well. In my case I wanted to specify all possible labels for the scikit-learn OrdinalEncoder, and the possible labels were retrieved from the dataset, which was stored as numpy floats. Extending for numpy types should be relatively simple, given we can retrieve them programmatically:
```
>>> np.sctypes
{'int': [<class 'numpy.int8'>, <class 'numpy.int16'>, <class 'numpy.int32'>, <class 'numpy.int64'>],
'uint': [<class 'numpy.uint8'>, <class 'numpy.uint16'>, <class 'numpy.uint32'>, <class 'numpy.uint64'>],
'float': [<class 'numpy.float16'>, <class 'numpy.float32'>, <class 'numpy.float64'>],
'complex': [<class 'numpy.complex64'>, <class 'numpy.complex128'>],
'others': [<class 'bool'>, <class 'object'>, <class 'bytes'>, <class 'str'>, <class 'numpy.void'>]}
```
e.g.
```
simple_numpy_types = [nptype for type_cat, nptypes in np.sctypes.items()
                             for nptype in nptypes
                             if type_cat != 'others']
simple_types = tuple([bool, int, float, str] + simple_numpy_types])
```
and then
`elif isinstance(o, simple_types) or o is None:`
I expected this to work, **but it does not**.
```
... traceback ...
  File "D:\repositories/openml-python\openml\flows\sklearn_converter.py", line 534, in _extract_information_from_model
    raise TypeError(msg)
TypeError: Second item of tuple does not match assumptions. Expected OpenMLFlow, got <class 'numpy.float32'>
```
I also tested just casting the numpy types to python native types but that did not work either. It looks like fixing this might need a deeper dive into the code. Nevertheless, I hope any of this helped whomever ends up working on this.

**Comment by PGijsbers** [comment]
First off, a (artificial) minimal example that exposes the issue:
```
import numpy as np
import openml
from sklearn.tree import DecisionTreeClassifier

openml.flows.sklearn_to_flow(DecisionTreeClassifier(max_depth=np.float32(3.0))
```
this currently raises
```
TypeError: (3.0, <class 'numpy.float32'>)
```
After patching in the `simple_types` check of the previous comment, the following error is raised instead:
```
TypeError: Object of type 'float32' is not JSON serializable
```
So we could extend the `sklearn_to_flow` by casting to the native data types:
```
elif isinstance(o, simple_types) or o is None:
    if type(o) in simple_numpy_types:
        o = o.item()
    # base parameter values
    rval = o
```
This seems to work as expected.

**Comment by mfeurer** [comment]
It appears like this is a common feature request. The only thing I'm a bit worried about is that we cannot easily do the de-serialization and get a numpy dtype again. Maybe that is not an issue for simple scalars but only for arrays?

**Comment by PGijsbers** [comment]
For the types that would now be included for serialization (various bit-length ints, uints, floats and complex), I don't foresee an issue. Could you come up with a concrete issue where it would matter? Otherwise I think it it's smarter to add this feature and tackle the problem if it actually comes up.

**Comment by mfeurer** [comment]
Makes sense. Please assign me to your PR once it is ready.

**Comment by mfeurer** [comment]
Closed via #635.
